### PR TITLE
crimson/osd/recovery_backend: restart object pulls that are blocked by down osds

### PR DESCRIPTION
**Pull request by xxhdx1985126**
Fixes: https://tracker.ceph.com/issues/67508
Signed-off-by: Xuehan Xu <xuxuehan@qianxin.com>

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

File: src/crimson/osd/recovery_backend.h [hunk 1]
```diff
@@ -219,6 +226,13 @@ class RecoveryBackend {
 	pulled.reset();
       }
     }
+    void repeat_pull() {
+      ceph_assert(pulled);
+      pulled->set_exception(crimson::ct_error::eagain::exception_ptr());
+    }
+    bool is_pulling() const {
+      return (bool)pulled;
```
Comment (Matan-B) [inline src/crimson/osd/recovery_backend.h:234]: nit: pulled.has_value() to avoid casting

**Review by Matan-B — APPROVED** [review]
lgtm, should this be tested with https://github.com/ceph/ceph/pull/58136?

File: src/crimson/osd/pg.h [hunk 2]
```diff
@@ -357,7 +357,13 @@ class PG : public boost::intrusive_ref_counter<
     shard_services.remove_want_pg_temp(orderer, pgid.pgid);
   }
   void check_recovery_sources(const OSDMapRef& newmap) final {
-    // Not needed yet
+    recovery_backend->for_each_recovery_waiter(
+      [newmap](auto &, auto &waiter) {
+        if (waiter->is_pulling() &&
+            newmap->is_down(waiter->pull_info->from.osd)) {
+          waiter->repeat_pull();
```
[the remark this answers is not in the document]
Comment (Matan-B) [reply src/crimson/osd/pg.h:364]: Please mark as needs-crimson-qa when ready, thanks :)

File: src/crimson/osd/replicated_recovery_backend.cc [hunk 3]
```diff
@@ -113,28 +113,30 @@ ReplicatedRecoveryBackend::maybe_pull_missing_obj(
...
+      return shard_services.send_to_osd(
+        pull_info.from.osd,
+        std::move(msg),
+        pg.get_osdmap_epoch());
+    }).si_then([this, soid] {
+      auto& recovery_waiter = get_recovering(soid);
+      return recovery_waiter.wait_for_pull();
+    });
   }).handle_error_interruptible(
     crimson::ct_error::assert_all("unexpected error")
```
[the remark this answers is not in the document]
Comment (xxhdx1985126) [reply src/crimson/osd/replicated_recovery_backend.cc:141]: In our tests, the recovery got blocked waiting for pull responses.

**Comment by xxhdx1985126** [comment]
> lgtm, should this be tested with #58136?

Yes

**Comment by xxhdx1985126** [comment]
Just fixed a bug in this PR, please take a look again.

File: src/crimson/osd/pg.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (Matan-B) [reply src/crimson/osd/pg.h:364]: @xxhdx1985126

[the remark this answers is not in the document]
Comment (xxhdx1985126) [reply src/crimson/osd/pg.h:364]: Ah, sorry, missed it, will add

**Comment by Matan-B** [comment]
https://pulpito.ceph.com/matan-2024-08-26_07:29:24-crimson-rados-wip-matanb-crimson-testing-aug-22-distro-crimson-smithi/

**Comment by Matan-B** [comment]
```
/home/jenkins-build/build/workspace/ceph-pull-requests/src/crimson/osd/pg.h:368:14: error: 'this' cannot be implicitly captured in this context

/home/jenkins-build/build/workspace/ceph-pull-requests/src/crimson/osd/pg.h:362:21: note: explicitly capture 'this'
      [newmap, FNAME](auto &, auto &waiter) {
                    ^
                    , this

```

**Comment by xxhdx1985126** [comment]
> ```
> /home/jenkins-build/build/workspace/ceph-pull-requests/src/crimson/osd/pg.h:368:14: error: 'this' cannot be implicitly captured in this context
> 
> /home/jenkins-build/build/workspace/ceph-pull-requests/src/crimson/osd/pg.h:362:21: note: explicitly capture 'this'
>       [newmap, FNAME](auto &, auto &waiter) {
>                     ^
>                     , this
> ```

Done:-)

**Comment by Matan-B** [comment]
https://pulpito.ceph.com/matan-2024-08-26_07:29:24-crimson-rados-wip-matanb-crimson-testing-aug-22-distro-crimson-smithi/

**Pull request merged**